pr_Homework_1

RegisEmiel opened the pull request.

RegisEmiel committed:
Инициализация ветки домашнего задания №1

RegisEmiel committed:
Инициализация ветки домашнего задания №1

RegisEmiel committed:
Домашнее задание №1 к сдаче

stvort commented on homework-1/src/main/java/otus/spring/homework_1.java:
```diff
@@ -0,0 +1,20 @@
+import otus.spring.service.ExaminerServiceImpl;
+
+
+public class homework_1 {
```
Классы в Java именуются в CamelCase

stvort commented on homework-1/pom.xml:
```diff
@@ -0,0 +1,54 @@
+    </dependencies>
+
+    <build>
+        <plugins>
```
О чем бы предупредил плагин checkstyle, который мы договаривались подключать к проектам

stvort commented on homework-1/src/main/java/otus/spring/homework_1.java:
```diff
@@ -0,0 +1,20 @@
+
+public class homework_1 {
+    public static void main(String[] args) {
+        System.out.println("Hello world");
```
В `main` не должно быть бизнес-логики кроме сборки старта приложения. Эту роль выполняет `examinerService.runTest();`, остальное лучше делать внутри

stvort commented on homework-1/src/main/java/otus/spring/service/ExaminerServiceImpl.java:
```diff
@@ -0,0 +1,24 @@
+@NoArgsConstructor
+@AllArgsConstructor
+public class ExaminerServiceImpl implements ExaminerService {
+    private QuestionRepository questionRepository;
```
`final`

stvort commented on homework-1/src/main/java/otus/spring/service/ExaminerServiceImpl.java:
```diff
@@ -0,0 +1,24 @@
+    public void runTest() {
+        List<Question> questions = questionRepository.getAll();
+
+        questions.forEach(System.out::println);
```
Работу с консолью (ввод/вывод, `System.in/out`) можно вынести в отдельный сервис. Так будет удобнее тестировать классы, где этот ввод/вывод используется. Причем `System.in/out` такому сервису можно будет тоже передать через конструктор. Такой сервис ничего не будет знать о предметной области (будет выводить и читать только строки и числа) и может быть переиспользован

stvort commented on homework-1/src/main/java/otus/spring/domain/Question.java:
```diff
@@ -0,0 +1,52 @@
+@NoArgsConstructor
+@AllArgsConstructor
+public class Question {
+    @CsvBindByName(required = true)
```
Бизнес сущности лучше не размечать аннотациями сериализации/десериализации. Это лучше делать с DTO

stvort commented on homework-1/src/main/java/otus/spring/domain/Question.java:
```diff
@@ -0,0 +1,52 @@
+    @CsvBindByName(required = true)
+    private String questionText;
+
+    @CsvBindAndSplitByName(elementType = String.class, collectionType = ArrayList.class, splitOn = "\\,", required = true)
+    private List<String> answers;
+
+    @CsvBindAndSplitByName(elementType = Integer.class, collectionType = ArrayList.class, splitOn = "\\,", required = true)
+    private  List<Integer> rightNumbers;
```
Лучше список объектов `Answer`, каждый из которых может быть правильным или нет. Ну и по заданию
> отдаём предпочтение объектам и классам, а не строчкам и массивам/спискам строчек

stvort commented on homework-1/src/main/java/otus/spring/domain/Question.java:
```diff
@@ -0,0 +1,52 @@
+    @CsvBindAndSplitByName(elementType = Integer.class, collectionType = ArrayList.class, splitOn = "\\,", required = true)
+    private  List<Integer> rightNumbers;
+
+    @Override
+    public String toString() {
+
+        StringBuilder sb = new StringBuilder("Вопрос №");
+        sb.append(number);
+        sb.append(System.lineSeparator());
+        sb.append(questionText);
+        sb.append(System.lineSeparator());
+
+        int answerNumber = 0;
+        for (String answer: answers) {
+            sb.append(++answerNumber).append(". ");
+            sb.append(answer);
+
+            if (rightNumbers.contains(answerNumber))
+                sb.append("\t*");
+
+            sb.append(System.lineSeparator());
+        }
+
+        return  sb.toString();
+    }
```
`toString` лучше не использовать для формирования пользовательского вывода. Вариантов представления сущности может быть много, особенно с учетом того, что она в теории вообще может использоваться в разных приложениях

RegisEmiel committed:
Доработка 1
Добавлен dto, тест, класс Answer
Исправлено название классов
Переработан класс вопросов

RegisEmiel committed:
Доработка 1.1
Добавлены тесты, сервис вывода
Проверка Checkstyle

stvort commented on homework-1/src/main/java/otus/spring/service/ExaminerServiceImpl.java:
```diff
@@ -0,0 +1,27 @@
+
+import java.util.List;
+
+@Getter
+@Setter
+@AllArgsConstructor
```
Лучше только `@RequiredArgsConstructor`

stvort commented on homework-1/src/main/java/otus/spring/domain/Question.java:
```diff
@@ -0,0 +1,40 @@
+        this.answers = questionDTO.getAnswers();
+    }
+
+    public String getStringRepresentation(boolean showCorrectAnswer) {
```
Преобразование себя в строку для вывода на UI не является ответственностью доменных объектов. Для этого лучше иметь отдельный конвертер используемый снаружи доменного класса

stvort commented on homework-1/src/main/java/otus/spring/domain/Question.java:
```diff
@@ -0,0 +1,40 @@
+
+    private List<Answer> answers;
+
+    public Question(QuestionDTO questionDTO) {
+        this.number = questionDTO.getNumber();
+        this.questionText = questionDTO.getQuestionText();
+        this.answers = questionDTO.getAnswers();
+    }
```
Создание себя из DTO тоже не является ответственностью доменных объектов. DTO может быть много, а доменный класс один. Иногда на несколько приложений. Добавление конвертаций в логику доменного класса нарушает OCP из SOLID

stvort commented on homework-1/src/main/java/otus/spring/dao/QuestionRepositoryImpl.java:
```diff
@@ -0,0 +1,63 @@
+@AllArgsConstructor
+public class QuestionRepositoryImpl implements QuestionRepository {
+
+    private String fileName;
+
+    private char separator;
```
`final`

stvort commented on homework-1/src/main/java/otus/spring/dao/QuestionRepositoryImpl.java:
```diff
@@ -0,0 +1,63 @@
+import java.io.Reader;
+import java.util.List;
+
+@Setter
+@Getter
+@NoArgsConstructor
+@AllArgsConstructor
```
`@RequiredArgsConstructor`

stvort commented on homework-1/src/main/java/otus/spring/dao/QuestionRepositoryImpl.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    private Reader getFileReader(String fileName) {
+        InputStream inputStream = getClass().getResourceAsStream(fileName);
```
Где закрывается данный стрим?

stvort commented on homework-1/src/main/java/otus/spring/service/QuestionConsoleOutputService.java:
```diff
@@ -0,0 +1,19 @@
+    private final PrintStream printStream = System.out;
+
+    @Override
+    public void outQuestion(String st) {
```
А строку, которая содержит не вопрос, а что-то другое, получается нельзя выводить? В занятии по архитектуре есть пример сервиса ввода-вывода. Рекомендую использовать его. Кстати там же, обсуждались предпосылки его появления)

stvort commented on homework-1/src/test/java/otus/spring/dao/QuestionRepositoryTest.java:
```diff
@@ -0,0 +1,42 @@
+
+@DisplayName("Тесты доступа к репозиторию вопросов")
+class QuestionRepositoryTest {
+    static private int QUESTION_COUNT = 2;
```
`private static final int QUESTION_COUNT = 2;`

RegisEmiel committed:
Доработка 2
Исправлены аннотации
Исправлен синтаксис
Исправлены наименования Output сервиса
Преобразование вопросов в строковое представление вынесено в отдельный класс
Создание вопросов из dto вынесено в отдельный класс

stvort commented on homework-1/src/main/java/otus/spring/service/ConsoleOutputService.java:
```diff
@@ -0,0 +1,5 @@
+package otus.spring.service;
+
+public interface ConsoleOutputService {
```
Задача интерфейса, скрывать конкретную реализацию, что-то у чего есть в названии `Console` может работать только с ней. `OutputService` же, подойдет для более широкого количества реализаций. И еще раз:
>В занятии по архитектуре есть пример сервиса ввода-вывода. Рекомендую использовать его. Кстати там же, обсуждались предпосылки его появления)

Давайте все же посмотрим, что там был за сервис и почему он был именно такой, а не какой-то другой

stvort commented on homework-1/src/main/java/otus/spring/service/ExaminerServiceImpl.java:
```diff
@@ -0,0 +1,32 @@
+
+    private final boolean fShowAnswer;
+
+    private final QuestionToStringConverter questionToStringConverter = new QuestionToStringConverter();
```
>3. Все зависимости должны быть настроены в IoC контейнере.

Это зависимость. Ее тоже надо сделать компонентом и внедрить. Руками мы не создаем компоненты

stvort commented on homework-1/src/main/java/otus/spring/utilities/CsvAnswerConverter.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    @Override
+    public Object convertToRead(String s) throws CsvDataTypeMismatchException, CsvConstraintViolationException {
```
Тут не кидается ни одного из этих исключений

stvort commented on homework-1/src/main/java/otus/spring/utilities/CsvAnswerConverter.java:
```diff
@@ -0,0 +1,28 @@
+
+public class CsvAnswerConverter extends AbstractCsvConverter {
+    @Override
+    public String convertToWrite(Object value) throws CsvDataTypeMismatchException {
+        Answer answer = (Answer) value;
```
Если мы все равно приводим к `Answer`, то и в метод нужно чтобы передавался он. Для этого можно использовать дженерики

stvort commented on homework-1/src/main/java/otus/spring/dao/QuestionRepositoryImpl.java:
```diff
@@ -0,0 +1,56 @@
+
+    private List<QuestionDTO> getQuestionsFromFile() {
+
+        InputStream inputStream = getClass().getResourceAsStream(fileName);
```
>Где закрывается данный стрим?

Стрим все еще не закрывается

RegisEmiel committed:
Доработка 3
Обернул потоки в try с ресурсами
Исправил наименование интерфейса OutputService
Убрал лишние исключения в CsvAnswerConverter
Убрал convertToWrite в CsvAnswerConverter, мы ничего не записываем
Сделал QuestionToStringConverter бином

stvort commented on homework-1/src/main/java/otus/spring/dto/QuestionDTOListConverter.java:
```diff
@@ -0,0 +1,26 @@
+
+import java.util.List;
+
+public class QuestionDTOListConverter {
```
Это тоже лучше сделать бином
